### PR TITLE
BUG-26214 fix colors parameter in DataFrame.boxplot

### LINKED ISSUE #26214: _dataframe.boxplot_ with _where_ and _by_ does not respect color keyword
### Bug report

**Bug summary**

The boxplot method on a dataframe which is using the "column, by" keywords does
not respect the _color_ keyword, and in fact crashes if it is present. This is not consistent with the documentation [here](http://pandas.pydata.org/pandas-docs/stable/user_guide/visualization.html#box-plots).

**Code for reproduction**

```python
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

def make_dummy_data():
    """ Return """
    df1 = pd.DataFrame(np.random.rand(10, 3), columns = ['x', 'y', 'z'])
    df2 = pd.DataFrame(2*np.random.rand(10, 3), columns = ['x', 'y', 'z'])
    return df1, df2

def comparative_results():
    """ stuff """

    df1, df2 = make_dummy_data()

    def draw_plot(ax, data, edge_color, fill_color=None):
        """ Controls details of color"""
        colors = dict(boxes=edge_color, whiskers=edge_color, medians=edge_color, caps=edge_color)
        ax = data.boxplot(column=['x'], by=['z'], showfliers=False, ax=ax, color=colors)
        return ax

    ax = None
    ax = draw_plot(ax, df1, 'k')
    ax = draw_plot(ax, df2, 'r')
    ax.set_title('dummy to expose bug')
    plt.show()


if __name__ == "__main__":
    comparative_results() 
```

**Actual outcome**

```
Traceback (most recent call last):
  File "/Users/BNL28/Code/DataPerformance/bug_report.py", line 33, in <module>
    comparative_results()
  File "/Users/BNL28/Code/DataPerformance/bug_report.py", line 26, in comparative_results
    ax = draw_plot(ax, df1, 'k')
  File "/Users/BNL28/Code/DataPerformance/bug_report.py", line 22, in draw_plot
    ax = data.boxplot(column=['x'], by=['z'], showfliers=False, ax=ax, color=colors)
  File "/Users/BNL28/anaconda3/lib/python3.6/site-packages/pandas/plotting/_core.py", line 2254, in boxplot_frame
    return_type=return_type, **kwds)
  File "/Users/BNL28/anaconda3/lib/python3.6/site-packages/pandas/plotting/_core.py", line 2223, in boxplot
    return_type=return_type)
  File "/Users/BNL28/anaconda3/lib/python3.6/site-packages/pandas/plotting/_core.py", line 2683, in _grouped_plot_by_column
    re_plotf = plotf(keys, values, ax, **kwargs)
  File "/Users/BNL28/anaconda3/lib/python3.6/site-packages/pandas/plotting/_core.py", line 2191, in plot_group
    bp = ax.boxplot(values, **kwds)
  File "/Users/BNL28/anaconda3/lib/python3.6/site-packages/matplotlib/__init__.py", line 1810, in inner
    return func(ax, *args, **kwargs)
TypeError: boxplot() got an unexpected keyword argument 'color'

Process finished with exit code 1 
```

**Expected outcome**

Expect two sets of box plots, one coloured black, and one coloured red. Code runs ok with no color keyword, but the boxes are indistinguishable without colour control.

**Environment**

  * Operating system: OSX
  * Matplotlib version: 3.0.2
  * Matplotlib backend (`print(matplotlib.get_backend())`):
  * Python version: Python 3.6.8 |Anaconda, Inc.| (default, Dec 29 2018, 19:04:46) 
  * Pandas version 0.24.2

### PR DESCRIPTION
**Pull request by JustinZhengBC**
- [X] closes #26214
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

This PR fixes issues in the handling of the `color` parameter in the implementation of `DataFrame.boxplot` that do not exist in `DataFrame.plot.box`

- the `color` parameter is now removed from `kwds` before the `matplotlib` call, preventing an error
- an alternative implementation of color extraction from `color`, as `_get_standard_colors` did not preserve the appropriate order (without looking at that function, one can see that the previous implementation would give the boxes and the medians the same color even when otherwise specified)
- checking that `color` is actually a `dict`

### CODE REVIEW

File: pandas/plotting/_core.py
Comment (WillAyd) [inline pandas/plotting/_core.py]: Not super familiar with this code but at first glance this seems kind of arbitrary - what's the reason for the selection here?

File: doc/source/whatsnew/v0.25.0.rst
Comment (WillAyd) [inline doc/source/whatsnew/v0.25.0.rst]: Does the behavior here match exactly what you'd get with DataFrame.plot.box? I assume the latter just passes kwargs on through but this looks to do special handling, so not entirely equivalent right?

Comment (JustinZhengBC) [reply doc/source/whatsnew/v0.25.0.rst]: `DataFrame.plot.box` does special handling too. Neither `color` nor any of its valid keys are listed as a valid kwarg in [matplotlib.pyplot.boxplot](https://matplotlib.org/api/_as_gen/matplotlib.pyplot.boxplot.html)

File: pandas/plotting/_core.py
Comment (JustinZhengBC) [reply pandas/plotting/_core.py]: I was trying to preserve previous behaviour for cases where `color` is not passed. If that's not required, then we can definitely change the color scheme generation to be simpler or even remove it entirely (falling back on `matplotlib` defaults)

Comment (WillAyd) [reply pandas/plotting/_core.py]: Ah OK I see - thanks for clarifying

**Comment by WillAyd** [comment]
So does `plot.box` accept this and are they now fully aligned? cc @datapythonista

**Comment by datapythonista** [comment]
Would be nice to wait until #26414 is in master to merge this.

**Comment by jreback** [comment]
we merged the refactor of the mpl backend, can you merge master and update to the new locations.

File: pandas/plotting/_matplotlib/boxplot.py
Comment (TomAugspurger) [inline pandas/plotting/_matplotlib/boxplot.py]: Should invalid colors raise? To catch issues like `.boxplot(colors={'boxse': 'red'})`

Comment (JustinZhengBC) [reply pandas/plotting/_matplotlib/boxplot.py]: They did in `DataFrame.plot.box`, so yes, I think so

**Comment by JustinZhengBC** [comment]
@jreback I've merged master and updated the PR

File: pandas/plotting/_matplotlib/boxplot.py
Comment (jreback) [inline pandas/plotting/_matplotlib/boxplot.py]: could be is_dict_like

File: pandas/plotting/_matplotlib/boxplot.py
Comment (jreback) [inline pandas/plotting/_matplotlib/boxplot.py]: can you add a bit more comments on what you are trying to achieve here?

maybe best to just put a more full doc-string at the top

**Comment by WillAyd** [comment]
@JustinZhengBC can you move whatsnew to 1.0.0 and merge master?

**Review by WillAyd — COMMENTED** [review]
Just so I'm clear - is the logic here the same for DataFrame.plot.box or do they diverge?

File: pandas/plotting/_matplotlib/boxplot.py
Comment (WillAyd) [reply pandas/plotting/_matplotlib/boxplot.py]: Still somewhat confused by this. If it's a string wouldn't the `.pop` call two lines up fail before this ever even hits the `is_dict_like` check?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/plotting/_matplotlib/boxplot.py
Comment (JustinZhengBC) [reply pandas/plotting/_matplotlib/boxplot.py]: For simplicity, let's say the only kwds that are passed to boxplot are `color` and `vert=True`.

Suppose `color` is a string, like `"r"`. Then kwds will be `{color: "r", vert: True}` and popping its value at the key "color" will yield `"r"` and assign it to `color`

Suppose `color` is a dict, like `{boxes: "r", whiskers: "b"}`. Then kwds will be `{color: {boxes: "r", whiskers: "b"}, vert: True`, or a dict where one of its values is another dict. Then, popping its value at the key "color" will yield the dict `{boxes: "r", whiskers: "b"}` and assign it to `color`

Comment (WillAyd) [reply pandas/plotting/_matplotlib/boxplot.py]: Ah OK thanks for explaining. I was getting tripped up between `color` and `colors` but see now

**Review by WillAyd — APPROVED** [review]
looks like a merge conflict in whatsnew but otherwise lgtm. @TomAugspurger maybe thoughts on this?

**Comment by WillAyd** [comment]
@JustinZhengBC can you merge master and fix merge conflict?

@TomAugspurger any thoughts on this? If not planning to merge in 2 days

**Review by TomAugspurger — APPROVED** [review]

**Comment by TomAugspurger** [comment]
Fixed the merge conflict. Merge on green.

**Comment by WillAyd** [comment]
Thanks @JustinZhengBC !

**Pull request merged**
